File: backend/src/services/spoonacularService.js
const env = require("../config/env");
const { spoonacularClient } = require("../config/spoonacularClient");

function mapSpoonacularError(error) {
  const status = error.response?.status;

  if (status === 401 || status === 402) {
    const friendly = new Error("Spoonacular key is invalid or has reached its plan limit.");
    friendly.statusCode = 401;
    throw friendly;
  }

  if (status === 429) {
    const friendly = new Error("Spoonacular rate limit reached. Please try again later.");
    friendly.statusCode = 429;
    throw friendly;
  }

  if (status) {
    const friendly = new Error(`Spoonacular request failed with status ${status}.`);
    friendly.statusCode = 502;
    throw friendly;
  }

  const friendly = new Error("Unable to reach Spoonacular API.");
  friendly.statusCode = 503;
  throw friendly;
}

function assertSpoonacularKey() {
  if (!env.spoonacularApiKey) {
    const error = new Error("SPOONACULAR_API_KEY is not configured.");
    error.statusCode = 500;
    throw error;
  }
}

async function searchRecipesByIngredients({
  ingredients,
  number = 10,
  ranking = 1,
  ignorePantry = true
}) {
  assertSpoonacularKey();

  if (!ingredients || !ingredients.trim()) {
    const error = new Error("Ingredients are required (comma-separated).");
    error.statusCode = 400;
    throw error;
  }

  try {
    const { data } = await spoonacularClient.get("/recipes/findByIngredients", {
      params: {
        apiKey: env.spoonacularApiKey,
        ingredients: ingredients.trim(),
        number,
        ranking,
        ignorePantry
      }
    });

    const results = Array.isArray(data) ? data : [];

    return {
      totalResults: results.length,
      results: results.map((item) => ({
        recipeId: item.id,
        recipeName: item.title,
        recipeImageUrl: item.image,
        usedIngredientCount: item.usedIngredientCount ?? 0,
        missedIngredientCount: item.missedIngredientCount ?? 0,
        missedIngredients: (item.missedIngredients || []).map((ingredient) => ingredient.name)
      }))
    };
  } catch (error) {
    mapSpoonacularError(error);
  }
}

module.exports = {
  searchRecipesByIngredients
};

File: backend/services/spoonacular_service.py
from __future__ import annotations

from typing import Any, Dict, List

import requests
from fastapi import HTTPException

from app.config import settings


def _raise_spoonacular_error(resp: requests.Response) -> None:
    if resp.status_code in (401, 402):
        raise HTTPException(status_code=401, detail="Spoonacular key is invalid or quota is exhausted.")
    if resp.status_code == 429:
        raise HTTPException(status_code=429, detail="Spoonacular rate limit reached. Try again later.")
    raise HTTPException(status_code=502, detail=f"Spoonacular request failed with status {resp.status_code}.")


def search_recipes_by_ingredients(
    ingredients: str,
    number: int = 10,
    ranking: int = 1,
    ignore_pantry: bool = True,
) -> Dict[str, Any]:
    if not settings.spoonacular_api_key:
        raise HTTPException(status_code=500, detail="SPOONACULAR_API_KEY is not configured.")

    if not ingredients.strip():
        raise HTTPException(status_code=400, detail="ingredients is required (comma-separated).")

    url = f"{settings.spoonacular_base_url}/recipes/findByIngredients"
    params = {
        "apiKey": settings.spoonacular_api_key,
        "ingredients": ingredients.strip(),
        "number": number,
        "ranking": ranking,
        "ignorePantry": str(ignore_pantry).lower(),
    }

    try:
        resp = requests.get(url, params=params, timeout=12)
    except requests.RequestException as exc:
        raise HTTPException(status_code=503, detail=f"Unable to reach Spoonacular API: {exc}") from exc

    if resp.status_code >= 400:
        _raise_spoonacular_error(resp)

    data = resp.json()
    results: List[Dict[str, Any]] = data if isinstance(data, list) else []

    mapped = [
        {
            "recipeId": item.get("id"),
            "recipeName": item.get("title"),
            "recipeImageUrl": item.get("image"),
            "usedIngredientCount": item.get("usedIngredientCount", 0),
            "missedIngredientCount": item.get("missedIngredientCount", 0),
            "missedIngredients": [ing.get("name") for ing in item.get("missedIngredients", [])],
        }
        for item in results
    ]

    return {"totalResults": len(mapped), "results": mapped}
